### ISSUE TITLE
WebExtension

### ISSUE THREAD

**Issue by xmorave2** [edabg/jsprintsetup#39]
Hello, 
as you probably know, Firefox is going to make changes in extension system, so only webextension could run in firefox - is it even possible to make jsprintsetup as webextension?

https://blog.mozilla.org/addons/2017/02/16/the-road-to-firefox-57-compatibility-milestones/

**Comment by eleclerc85** [comment]
I'm looking foward to this me too!

**Comment by mitkola** [comment]
To implement jsPrintSetup as Webextension there is need for extending Webextension API.
I have made propose for new API and I'm waiting for review.
https://github.com/edabg/web-ext-experiment-printservice
https://bugzilla.mozilla.org/show_bug.cgi?id=1385827
With current  API, there is no possible implementation of jsPrintSetup as Webextension.

**Comment by xmorave2** [comment]
Thank you @mitkola for your work! Hope Mozilla will add neede API soon, Great job!

**Comment by eleclerc85** [comment]
Thx to mitkola for giving his time to the community.

**Comment by marcobottignoleatlas** [comment]
Sorry for my insistence, the development of webextension will or will not be known? I wonder why otherwise we have to think of another solution.
Sorry for my English, but Google is helping me ;-)

**Comment by mitkola** [comment]
We are waiting mentor to help us in landing proposed changes in FF first.
https://bugzilla.mozilla.org/show_bug.cgi?id=1385827
After this jsPrintSetup WE will be done.

**Comment by MacGritsch** [comment]
Feel free to add a comment on https://bugzilla.mozilla.org/show_bug.cgi?id=1385827 so that they see that many people need that feature.

Thank you!

**Comment by gainup** [comment]
That's all. They blocked it.

**Comment by NtH0** [comment]
> That's all. They blocked it.

The comments, not the ticket (yet).

**Comment by survtur** [comment]
Look's like i'll need to use older version of firefox....

**Comment by balping** [comment]
You can still vote on the issue.

**Comment by survtur** [comment]
@balping 
No, I can't.

> Restrict Comments: true
> You are not allowed to make an additional comment on this bug.

**Comment by balping** [comment]
Yes, you can. At the top, in the details box.

**Comment by MacGritsch** [comment]
I think that is the link for voting:

https://bugzilla.mozilla.org/page.cgi?id=voting/user.html&bug_id=1385827#vote_1385827

**Comment by mbarakaja** [comment]
Is unbelievable that Mozilla is starting to support more complex stuff like Virtual Reality, but they do not provide decent printer API. I now that the average browser users do not print often in the web, but a wide printing support should be an standard a long time ago.

Meanwhile, the [ticket](https://bugzilla.mozilla.org/page.cgi?id=voting/user.html&bug_id=1385827#vote_1385827) just have 9 votes so far. We need to make more noise people.

**Comment by gainup** [comment]
I don't think that this problem will be solved soon. Unfortunately I must find other solution. Almost sure will have to buy native application and rewrite a lot of code. Moreover, I doubt that disabling applications and copy everything from Chrome will increase Firefox market share.

**Comment by TomatoJoe** [comment]
@mbarakaja 

> I now that the average browser users do not print often in the web.

You may be surprised. Business users will be printing quite a lot, not only to paper for things like Postage labels and invoices, but to PDF for Tax records. 

> Is unbelievable that Mozilla is starting to support more complex stuff like Virtual Reality, but they do not provide decent printer API

As a business owner, I found that Mozilla, in particular, Firefox is the ONLY browser that could fulfil our requirements for printing. This assumption that "no body prints these days" is bollocks. And to take on what I call the "Apple mentality" of only supporting only the needs of the majority will see business users move away.  I too am astounded that there are NO web browsers out there that can accurately handle printing tasks without failing in at least one area. And that gos for operating systems equally. OSX is unusable in this regard.
We include code in our html pages to automatically pop up the Print Dialog box, it save so much time.

**Comment by J-N-K** [comment]
@gainup, I don't think this will be solved anytime. The last statement in the mozilla bug is quite clear: if you need something like that, go away.

**Comment by black23** [comment]
It can be solved by Native_messaging. What do You think @mitkola about it?
https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging

**Comment by balping** [comment]
Good luck on writing a program that renders html as firefox would and is able to create proper postscript/pdf files with headers and margins as required. Then good luck on providing and maintaining binaries for windows, mac and gnu/linux. And then good luck convincing clients why they should install a whole separate program just to have nice headers on a printed sheet.

It's certainly not impossible just a bit unnecessarily overcomplicated.

**Comment by mbarakaja** [comment]
@black23, as far I understand, **`JSPrintSetup`**, is not responsible of rendering the content to be sent to a given printer, this is still managed by the browser.

So, in order to implement features like silent printing (bypassing the browser implementation), the native application must have to manage the rendering by it self.

So the question is, **How do you send the content (a BLOB file, raw text, etc) using just Native messaging?** The Mozilla documentation says that just serializable JSON messages can be sent in and out between browser extension and a native application.

**Comment by MacGritsch** [comment]
Yea I have already developed an extension with native messaging and have tried to solve the printing problem this way. So first you need thir-party framework to render the html page and second you need a thir-party applicaton to print that pdf. And you have to install all that apps on the client machines (and place entries in the registry (facepalm)). And thats only for one platform - if you have to support all platforms on which FF runs then *good bye*
I think my clients stay on the ESR-version and when that is outdated, thy keep it and go online with another, save browser. So FF loses marketshare. If they want so...

**Comment by balping** [comment]
As far as I know, no other browser supports this feature, so ff doesn't necessarily loose market share, just the chance to stand out a bit. When ESR enters quantum (in March 2018), I think I'll switch to server-side pdf rendering.

**Comment by MacGritsch** [comment]
> I think I'll switch to server-side pdf rendering.

Thats also my plan, but then also an application is needed to print that PDfs :-/

**Comment by black23** [comment]
Thank You all for explanation about Native_messaging.
I've found "solution" thru Native_messaging on forum and it was propably directly from Mozilla guys. But I agree that can't solve our problem. I really don't uderstand Mozilla developers because they stoped debate about silence printig due security. It could be handled by user confirmation about printing without promt, for example.
ESR is just short time solution. PDF rendering isn't work best for every application, especially if you print just simple receipts.

**Comment by balping** [comment]
I came across [wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf). Maybe it could be used for the rendering part.

[It's possible ](https://wkhtmltopdf.org/usage/wkhtmltopdf.txt) to contol margins and header/footer text.

**Comment by kelemenbalazs** [comment]
Any solution so far? After quantum, you are not able to use the extension? Please, this is a too useful extension to throw it away. I would like to use, but I also have to take care of security, so I cannot stay at FF 56.0 because of the extension incompatibility.

**Comment by gainup** [comment]
> Any solution so far? After quantum, you are not able to use the extension? Please, this is a too useful extension to throw it away. I would like to use, but I also have to take care of security, so I cannot stay at FF 56.0 because of the extension incompatibility.

I've ordered native application on Windows for the printing purposes. It accepts data from the extension in json format, renders it and sends html page to printer.

**Comment by qfluxlab** [comment]
@gainup What native application are you working with? I'm interested in trying something similar.

**Comment by iwasingh** [comment]
Any updates? For now, the only solution is to use an older version of Firefox, am i right?

**Comment by mayankmrc3** [comment]
Any good news for JSprint?

**Comment by mben91** [comment]
Any updates about this issue? been using this for a considerate amount of time and I'll be happy not to change this. Using an older browser unfortunately is not an option for my case. Please let us know. Thanks

**Comment by gainup** [comment]
Only native application. If anyone is willing to do such for Windows, I would love to invest.

**Comment by optimistex** [comment]
@mohamedbenatti 
It works with: https://www.palemoon.org

**Comment by yeffer2** [comment]
https://www.palemoon.org mejor solucion ....
